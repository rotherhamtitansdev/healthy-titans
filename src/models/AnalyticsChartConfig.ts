import { NHSNutritionalDataModel } from "./NHSNutritionalDataModel";

// This is the numerical instance of the NHS Nutritional guidelines model
export const NHSNutritionalBoundaries: NHSNutritionalDataModel[] = [
  { name: "Fat", low: 3, high: 17.5 },
  { name: "Saturates", low: 1.5, high: 5 },
  { name: "Sugars", low: 5, high: 22.5 },
  { name: "Salt", low: 0.1, high: 0.6 },
];

// These are the config parameters for the chart styling and sizing
export const chartSizing = {
  chartHeight: 500,
  barWidth: 30,
  CStartPos: 50,
  RStartPos: 60,
  circleYDif: 100,
  smallerRounded: 15,
};

export enum ChartLabel {
  Calories = "ENERC_KCAL",
  Protein = "PROCNT",
  Fat = "FAT",
  Saturates = "FASAT",
  Sugars = "SUGAR",
  Salt = "NA",
}

export enum Colour {
  Green = 1,
  Yellow = 2,
  Red = 3,
}
