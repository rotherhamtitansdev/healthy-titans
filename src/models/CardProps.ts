export interface CardProps {
  name: string;
  additionalStyling?: string;
  onClick?: () => void;
}
