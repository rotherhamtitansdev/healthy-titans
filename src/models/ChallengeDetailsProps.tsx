export interface ChallengeDetailsProps {
  name: string;
  description: string;
  img: string;
  rules: string[];
  equipment: string[];
  howToPlay?: string[];
}
