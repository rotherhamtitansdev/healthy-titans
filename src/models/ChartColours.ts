export interface ChartColours {
  name: string;
  primary: string;
  secondary: string;
  status: number;
  position: number;
}
