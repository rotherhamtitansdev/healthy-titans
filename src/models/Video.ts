import React from "react";

export interface Video {
  firebaseName: string;
  title: string;
  description?: string;
}

export interface DrilledVideoProps {
  setHidden: React.Dispatch<React.SetStateAction<boolean>>
  setModalClickedVideoData: React.Dispatch<React.SetStateAction<Video | undefined>>
}

export interface VideoModalProps {
  video: Video
  getHidden: boolean;
  setHidden: React.Dispatch<React.SetStateAction<boolean>>
}
