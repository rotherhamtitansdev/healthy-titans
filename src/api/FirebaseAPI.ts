import { getDownloadURL, getStorage, ref } from "firebase/storage";
import {
  collection,
  doc,
  getDoc,
  getDocs,
  query,
  where,
  documentId,
} from "firebase/firestore";
import { FoodDetailsProps } from "../data/nutritional_information/FoodDetailsComponentData";
import { fStore } from "../config/firebase-config";
import { NutritionalDetailsFirebaseProps } from "../models/NutritionDetailsComponentData";

interface SeeNextProps {
  key: number;
  name: string;
  path: string;
  img: string;
}

class FirebaseAPI {
  static fetchImages = async (firebaseName: string): Promise<string> => {
    const storage = getStorage();
    const starsRef = ref(storage, firebaseName);
    return getDownloadURL(starsRef);
  };

  static fetchAllImages = async (firebaseNames: string[]) =>
    Promise.all(firebaseNames.map((firebaseName) => FirebaseAPI.fetchImages(firebaseName)));

  static fetchFoodDetailsComponentsData = async (): Promise<
    FoodDetailsProps[] | undefined
  > => {
    const querySnapshot = await getDocs(collection(fStore, "FYPData"));
    if (!querySnapshot) return undefined;
     return querySnapshot.docs.map((each) => each.data() as FoodDetailsProps);
  };

  static fetchFoodDetailsSingle = async (name: string) => {
    const docRef = doc(fStore, "FYPData", name);
    const docSnap = await getDoc(docRef);
    if (docSnap) return docSnap.data() as FoodDetailsProps;
    return undefined;
  };

  static fetchFoodDetailsSeeNext = async (
    category: string,
    currentName: string
  ): Promise<{ cardData: FoodDetailsProps[]; paths: string[] }> => {
    const q = query(
      collection(fStore, "FYPData"),
      where("category", "==", category),
      where("name", "!=", currentName)
    );
    const docSnap = await getDocs(q);
    const docs = docSnap.docs.map((data) => {
      const newDoc = data.data();
      newDoc.path = data.id;
      return newDoc;
    });

    const shuffled = docs.sort(() => 0.5 - Math.random()).slice(0, 3);

    return {
      cardData: shuffled.map((value) => ({
        name: value.name,
        description: value.description,
        firebaseName: value.firebaseName,
        category: value.category,
        score: value.score,
        facts: value.facts,
      })),
      paths: shuffled.map((value) => value.path),
    };
  };

  static fetchNutritionData = async (name: string): Promise<NutritionalDetailsFirebaseProps> => {
    const querySnapshot = await getDocs(collection(fStore, "NutritionData", name, "Content"));

    const sorted = querySnapshot.docs.sort((a, b) => {
      if (a.data().order > b.data().order) {
        return 1;
      }
      if (b.data().order > a.data().order) {
        return -1;
      }
      return 0;
    });

    return sorted.map((doc2) => {
      const data = doc2.data();
      return Object.entries(data).map(([key,value]) => ({key,value}));
    });
  };

  static fetchNutritionSeeNext = async (currentName: string): Promise<SeeNextProps[]> => {
    const q = query(collection(fStore, "NutritionData"), where(documentId(), "!=", currentName));
    const querySnapshot = await getDocs(q);

    const docs = querySnapshot.docs.sort(() => 0.5 - Math.random()).slice(0, 3);

    return Promise.all(
      docs.map(async (sortedDoc, index) => {
        const dataDoc = await getDoc(doc(fStore, "NutritionData", sortedDoc.id, "Content", "Main"));

        const URI = await FirebaseAPI.fetchImages(dataDoc.data()?.firebaseName);

        return { key: index, name: dataDoc.data()?.name, path: docs[index].id, img: URI };
      })
    );
  };

  static fetchQuizData = async () => {
    const docRef = doc(fStore, "QuizData", "Quiz");
    const quizDoc = await getDoc(docRef);

    if (quizDoc.exists()) {
      return quizDoc.data();
    }
    return undefined;
  };

  static fetchAboutUsData = async () => {
    const docRef = doc(fStore, "AboutUsData", "Data");
    const aboutUsDoc = await getDoc(docRef);

    if (aboutUsDoc.exists()) {
      return aboutUsDoc.data();
    }
    return undefined;
  };

  static fetchChallengesData = async (currentName: string) => {
    const docRef = doc(fStore, "Challenges", currentName);
    const challengesDoc = await getDoc(docRef);

    if (challengesDoc.exists()) {
      return challengesDoc.data();
    }
    return undefined;
  };

  static fetchChallengesList = async () => {
    const querySnapshot = await getDocs(collection(fStore, "ChallengesData"));

    if (!querySnapshot) return undefined;
    return querySnapshot.docs.map((each) => each.data());
  };
}

export default FirebaseAPI;
