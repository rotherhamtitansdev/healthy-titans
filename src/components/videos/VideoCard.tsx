import React, { useCallback, useEffect, useState } from "react";
import FirebaseAPI from "../../api/FirebaseAPI";
import "../../App.css";
import { DrilledVideoProps, Video } from "../../models/Video";
import Card from "../shared/Card";

// This displays the video and its details
const VideoCard = (props: { video: Video, Actions: DrilledVideoProps }) => {
  const [getVideoURL, setVideoURL] = useState<string | undefined>();

  useEffect(() => {
    if (props.video.firebaseName !== undefined) {
      FirebaseAPI.fetchImages(props.video.firebaseName).then((URI) => setVideoURL(URI));
    }
  }, []);

  const handleClick = useCallback(() => {
    props.Actions.setHidden(false);
    props.Actions.setModalClickedVideoData(props.video);
  }, []);

  return (
    <Card card={{ name: props.video.title, onClick: handleClick }}>
      <div className="grid grid-cols-2 gap-2">
        {getVideoURL && props.video.description ? (
          // eslint-disable-next-line jsx-a11y/media-has-caption
          <video width="400">
            <source src={getVideoURL} />
          </video>
        )
          : null}
        <div className="pt-5 pb-3">
          <div className="block mt-1 text-md leading-tight text-gray-900">
            {props.video.description}
          </div>
        </div>
      </div>
    </Card>
  );
};

export default VideoCard;
