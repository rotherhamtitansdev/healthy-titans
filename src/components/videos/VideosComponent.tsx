import React, { useState } from "react";
import Carousel from "react-multi-carousel";
import { VideoData } from "../../data/VideoData";
import { Video } from "../../models/Video";
import AppHeader from "../app_header/AppHeader";
import MenuHeader from "../app_header/header/MenuHeader";
import Card from "../shared/Card";
import MenuTitle from "../shared/MenuTitle";
import VideoCard from "./VideoCard";
import VideoModal from "./VideoModal";

const VideosComponent = () => {
  const [getHidden, setHidden] = useState(true);
  const [modalClickedVideoData, setModalClickedVideoData] = useState<Video | undefined>();
  const [isMoving, setIsMoving] = useState(false);

  const responsive = {
    desktop: {
      breakpoint: { max: 3000, min: 1600 },
      items: 2,
      partialVisibilityGutter: 200,
    },
    smallDesktop: {
      breakpoint: { max: 1600, min: 1200 },
      items: 2,
      partialVisibilityGutter: 0,
    },
    tablet: {
      breakpoint: { max: 1200, min: 600 },
      items: 1,
      partialVisibilityGutter: 120,
    },
    mobile: {
      breakpoint: { max: 600, min: 464 },
      items: 1,
      partialVisibilityGutter: 150,
    },
    smallMobile: {
      breakpoint: { max: 464, min: 0 },
      items: 1,
      partialVisibilityGutter: 50,
    },
  };

  const videoData = VideoData.map((item, index) => (
    <VideoCard
      key={index}
      Actions={{ setHidden, setModalClickedVideoData }}
      video={item}
      disableOnClick={isMoving}
    />
  ));

  function getVideos() {
    let videos = videoData;
    // Fill carousel with placeholders if there are fewer than 3 videos
    while (videos.length < 3) {
      videos = videos.concat(
        <Card card={{ name: "video placeholder", additionalStyling: "max-h-[20rem] md:max-h-[28rem] max-w-[22.5rem] sm:max-w-[37.5rem]" }}>
          <div className="text-center py-[4.5rem] px-[5rem] xs:py-[5.75rem] xs:px-[7.25rem] sm:py-[9.35rem] sm:px-[5rem] md:py-[9rem] md:px-[15rem] lg:py-[11.5rem] lg:px-[10rem]">
            <div className="font-semibold text-4xl font-quicksand text-homepageHeaderText">
              More videos coming soon!
            </div>
          </div>
        </Card>,
      );
    }
    return videos;
  }

  return (
    <>
      <AppHeader>
        <MenuHeader title="Videos" body="Here you can watch a few short videos and hear from the Titans as you continue to learn and get to know even more about how to be strong and healthy." />
      </AppHeader>
      {modalClickedVideoData
        && (
          <VideoModal
            video={modalClickedVideoData}
            getHidden={getHidden}
            setHidden={setHidden}
          />
        )}
      <div className="pl-5 md:py-10 md:pl-10">
        <MenuTitle title="Watch and learn with these short videos" subtitle="Click on whichever video you'd like to watch to learn more and hear from some Titans!" />
        <Carousel
          responsive={responsive}
          partialVisible
          beforeChange={() => setIsMoving(true)}
          afterChange={() => setIsMoving(false)}
          itemClass="p-[1rem]"
          renderArrowsWhenDisabled
          arrows={getHidden}
        >
          {getVideos()}
        </Carousel>
      </div>
    </>
  );
};

export default VideosComponent;
