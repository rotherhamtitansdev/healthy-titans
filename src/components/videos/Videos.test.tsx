import { cleanup, render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import React from "react";
import { Route } from "react-router";
import FirebaseAPI from "../../api/FirebaseAPI";
import { Video } from "../../models/Video";
import RoutingTestWrapper from "../../tests/RoutingTestWrapper";
import VideoCard from "./VideoCard";
import VideosComponent from "./VideosComponent";

const mockVideoData = [
  {
    description: "VIDEO",
    firebaseName: "VIDEO.mp4",
    title: "VIDEO",
  },
];

afterEach(() => {
  cleanup();
});

test("navigate to videos page", () => {
  const videosRoute = "/Videos";
  jest.spyOn(FirebaseAPI, "fetchDataFromPath").mockResolvedValue(mockVideoData);

  const snapshot = render(
    <RoutingTestWrapper path={videosRoute}>
      <Route path="/Videos" element={<VideosComponent />} />
    </RoutingTestWrapper>
  );

  waitFor(() => {
    expect(snapshot).toMatchSnapshot();
  });
});

describe("Videos page: Render", () => {
  jest.mock("react", () => ({
    ...jest.requireActual("react"),
    useState: jest.fn(),
  }));

  jest.spyOn(React, "useEffect").mockImplementation((f) => f());
  jest.spyOn(FirebaseAPI, "fetchImages").mockResolvedValue(mockVideoData[0].firebaseName);

  const VideoProps: Video = {
    firebaseName: "firebase url",
    title: "video title",
    description: "video description",
  };

  it("should render the video component", () => {
    const snapshot = render(
      <VideoCard
        Actions={{ setHidden: jest.fn(), setModalClickedVideoData: jest.fn() }}
        video={VideoProps}
        disableOnClick={false}
      />
    );
    expect(snapshot).toMatchSnapshot();
  });

  it("should unhide video player when video clicked", async () => {
    const user = userEvent.setup();
    const setHiddenMock = jest.fn();
    render(
      <VideoCard
        Actions={{ setHidden: setHiddenMock, setModalClickedVideoData: jest.fn() }}
        video={VideoProps}
        disableOnClick={false}
      />
    );

    const video = screen.getByTestId(VideoProps.title);
    expect(video).toBeInTheDocument();
    await user.click(video);

    expect(setHiddenMock).toHaveBeenCalledWith(false);
  });
});
