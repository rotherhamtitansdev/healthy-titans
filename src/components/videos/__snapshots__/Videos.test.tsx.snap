// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Videos page: Render should render the video component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="bg-white font-quicksand table align-middle rounded-3xl shadow-lg w-[18rem] xs:w-[22.5rem] sm:w-[30rem] md:w-[37.5rem] hover:cursor-grab hover:-translate-y-0.5 hover:scale-105 duration-300"
        data-testid="video title"
        role="button"
        tabindex="0"
      >
        <div
          class="p-5"
        >
          <div
            class="pt-5 pb-3"
          >
            <div
              class="font-semibold text-2xl  text-homepageHeaderText"
            >
              video title
            </div>
            <div
              class="font-md text-xl  text-homepageHeaderText"
            >
              video description
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="bg-white font-quicksand table align-middle rounded-3xl shadow-lg w-[18rem] xs:w-[22.5rem] sm:w-[30rem] md:w-[37.5rem] hover:cursor-grab hover:-translate-y-0.5 hover:scale-105 duration-300"
      data-testid="video title"
      role="button"
      tabindex="0"
    >
      <div
        class="p-5"
      >
        <div
          class="pt-5 pb-3"
        >
          <div
            class="font-semibold text-2xl  text-homepageHeaderText"
          >
            video title
          </div>
          <div
            class="font-md text-xl  text-homepageHeaderText"
          >
            video description
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`navigate to videos page 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="bg-mobileWave sm:bg-wave bg-no-repeat pb-2 bg-span"
      >
        <div
          class="flex flex-wrap h-[22rem]"
        >
          <div
            class="mx-auto pt-[3rem] md:pt-[2rem]"
          >
            <h1
              class="text-titansDarkBlue font-bold text-center sm:mt-[1rem] text-3xl sm:text-5xl lg:text-6xl mt-8 pb-4"
              data-testid="header-title"
            >
              Videos
            </h1>
            <div
              class="font-medium text-homepageHeaderSubTitle text-sm md:text-base lg:text-xl text-center w-2/3 lg:w-[45rem] mx-auto"
            >
              Here you can watch a few short videos and hear from the Titans as you continue to learn and get to know even more about how to be strong and healthy.
            </div>
          </div>
        </div>
      </div>
      <div
        class="pl-5 md:py-10 md:pl-10"
      >
        <div
          class="md:px-5"
        >
          <div
            class="font-semibold text-[#243A73] text-2xl md:text-3xl lg:text-4xl py-5"
            data-testid="menutitle-title"
          >
            Watch and learn with these short videos
          </div>
          <div
            class="font-medium text-[#243A73] md:text-xl lg:text-xl md:pb-5"
          >
            Click on whichever video you'd like to watch to learn more and hear from some Titans!
          </div>
        </div>
        <div
          class="react-multi-carousel-list  "
          dir="ltr"
        >
          <ul
            class="react-multi-carousel-track "
            style="transition: none; overflow: unset; transform: translate3d(0px,0,0);"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="bg-mobileWave sm:bg-wave bg-no-repeat pb-2 bg-span"
    >
      <div
        class="flex flex-wrap h-[22rem]"
      >
        <div
          class="mx-auto pt-[3rem] md:pt-[2rem]"
        >
          <h1
            class="text-titansDarkBlue font-bold text-center sm:mt-[1rem] text-3xl sm:text-5xl lg:text-6xl mt-8 pb-4"
            data-testid="header-title"
          >
            Videos
          </h1>
          <div
            class="font-medium text-homepageHeaderSubTitle text-sm md:text-base lg:text-xl text-center w-2/3 lg:w-[45rem] mx-auto"
          >
            Here you can watch a few short videos and hear from the Titans as you continue to learn and get to know even more about how to be strong and healthy.
          </div>
        </div>
      </div>
    </div>
    <div
      class="pl-5 md:py-10 md:pl-10"
    >
      <div
        class="md:px-5"
      >
        <div
          class="font-semibold text-[#243A73] text-2xl md:text-3xl lg:text-4xl py-5"
          data-testid="menutitle-title"
        >
          Watch and learn with these short videos
        </div>
        <div
          class="font-medium text-[#243A73] md:text-xl lg:text-xl md:pb-5"
        >
          Click on whichever video you'd like to watch to learn more and hear from some Titans!
        </div>
      </div>
      <div
        class="react-multi-carousel-list  "
        dir="ltr"
      >
        <ul
          class="react-multi-carousel-track "
          style="transition: none; overflow: unset; transform: translate3d(0px,0,0);"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
